File: Algorithm/数组/移除元素/removeElement.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int removeElement(vector<int> &nums, int val)
    {
        int slowIndex, fastIndex = 0;
        for (int fastIndex = 0; fastIndex < nums.size(); fastIndex++)
        {
            if (nums[fastIndex] != val)
            {
                nums[slowIndex] = nums[fastIndex];
                slowIndex++;
            }
        }
        return slowIndex;
    }
    int removeElement2(vector<int> &nums, int vla)
    {
        int leftIndex = 0;
        int rightIndex = nums.size() - 1;
        while (leftIndex <= rightIndex)
        {
            while (leftIndex <= rightIndex && nums[leftIndex] != vla)
            {
                leftIndex++;
            }
            while (leftIndex <= rightIndex && nums[rightIndex] == vla)
            {
                rightIndex--;
            }
            if (leftIndex <= rightIndex)
            {
                nums[leftIndex++] = nums[rightIndex--];
            }
        }
        return leftIndex;
    };
};

int main()
{
    vector<int> nums = {1, 2, 3, 2, 4, 5, 2, 6};
    int val = 2;
    Solution s;
    int res = s.removeElement2(nums, val);
    cout << res << endl;
    return 0;
}

File: Algorithm/数组/移除元素/removeElement.py
class Solution:
    def removeElement1(self, nums: list[int], val: int):
        slowIndex = 0
        for fastIndex in range(len(nums)):
            if nums[slowIndex] != val:
                nums[slowIndex] = nums[fastIndex]
                slowIndex += 1
        return slowIndex
    
    def removeElement2(self, nums: list[int], val: int):
        leftIndex = 0
        rightIndex = len(nums) - 1
        while leftIndex <= rightIndex:
            while leftIndex <= rightIndex and nums[leftIndex] != val:
                leftIndex += 1
            while leftIndex <= rightIndex and nums[rightIndex] == val:
                rightIndex -= 1
            if leftIndex <= rightIndex:
                nums[leftIndex], nums[rightIndex] = nums[rightIndex], nums[leftIndex]
                leftIndex += 1
                rightIndex -= 1
        return leftIndex

if __name__ == "__main__":
    nums = [3,2,2,3]
    val = 3
    s = Solution()
    print(s.removeElement1(nums, val))
    print(nums)
    nums = [3,2,2,3]
    print(s.removeElement2(nums, val))
    print(nums)